### ISSUE TITLE
Many-To-Many relationship throws NotImplementedException

### ISSUE THREAD

**Issue by langebo** [dotnet/efcore#22241]
When configuring a Many-To-Many relationship between entities with Guid keys using a mapping entity I get a NotImplementedException with the message 'TODO: Other kinds of keys'. This might be a non-issue since #10508 is associated to the 5.0.0-rc1 milestone and this is tested with preview8 packages. 

Resource model
```csharp
public class Resource
{
    public Guid Id { get; set; }
    public string Title { get; set; }

    public virtual List<Role> Roles { get; set; }
    public virtual List<ShiftTrack> Tracks { get; set; }

    public Resource()
    {
        Roles = new List<Role>();
        Tracks = new List<ShiftTrack>();
    }
}
```

Role model
```csharp
public class Role
{
    public Guid Id { get; set; }
    public string Title { get; set; }
    public virtual List<Resource> Resources { get; set; }

    public Role()
    {
        Resources = new List<Resource>();
    }
}
```

ResourceRole mapping entity
```csharp
public class ResourceRole
{
    public Guid Id { get; set; }

    public Guid RoleId { get; set; }
    public virtual Role Role { get; set; }

    public Guid ResourceId { get; set; }
    public virtual Resource Resource { get; set; }
}
```

OnModelCreating within the DbContext
``` csharp
modelBuilder.Entity<Resource>(e =>
{
    e.HasMany(r => r.Roles)
     .WithMany(r => r.Resources)
     .UsingEntity<ResourceRole>(r => r.HasOne(e => e.Role).WithMany().HasForeignKey(e => e.RoleId),
                                l => l.HasOne(e => e.Resource).WithMany().HasForeignKey(e => e.ResourceId));
});
```

Code that tries to assign a role to a resource (both are not null)
``` csharp
var resource = await db.Resources
    .Include(r => r.Roles)
    .FirstOrDefaultAsync(r => r.Id == id);

var role = await db.Roles
    .FirstOrDefaultAsync(r => !resource.Roles.Contains(r));

resource.Roles.Add(role);

await db.SaveChangesAsync(); 
```

Results in this Exception + StackTrace on the SaveChangesAsync() call
```csharp
Exception thrown: 'System.NotImplementedException' in System.Private.CoreLib.dll: 'TODO: Other kinds of keys'
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.FindAssociationEntry(InternalEntityEntry entry, InternalEntityEntry otherEntry, ISkipNavigation skipNavigation)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.FindOrCreateAssociationEntry(InternalEntityEntry entry, InternalEntityEntry otherEntry, ISkipNavigation skipNavigation, Boolean fromQuery, Boolean setModified)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.NavigationCollectionChanged(InternalEntityEntry entry, INavigationBase navigationBase, IEnumerable`1 added, IEnumerable`1 removed)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryNotifier.NavigationCollectionChanged(InternalEntityEntry entry, INavigationBase navigationBase, IEnumerable`1 added, IEnumerable`1 removed)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectNavigationChange(InternalEntityEntry entry, INavigationBase navigationBase)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.LocalDetectChanges(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(IStateManager stateManager)
   at Microsoft.EntityFrameworkCore.ChangeTracking.ChangeTracker.DetectChanges()
   at Microsoft.EntityFrameworkCore.DbContext.TryDetectChanges()
   at Microsoft.EntityFrameworkCore.DbContext.<SaveChangesAsync>d__56.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at WorkF.App.Controllers.ResourcesController.<AddRole>d__5.MoveNext() in /Users/REDACTED/dev/priv/WorkF/WorkF.App/Controllers/ResourcesController.cs:line 68
```

### Further technical details

EF Core version: 5.0.0-preview.8.20407.4
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 5.0 (SDK 5.0.100-preview.8.20417.9)
Operating system: macOS Catalina 10.15.6 (19G2021)
IDE: Visual Studio Code 1.48.2

**Comment by ajcvickers** [comment]
Fixed in #21902 or related PR. I'm curious why you have a separate FK column in your join table. What is the advantage of doing this? (See also #21900 and #21901.)

**Comment by langebo** [comment]
@ajcvickers Cheers. In this case there is no advantage to be honest. Was just a little throwaway project to test all the good stuff of net5.0. So I probably was just quickly rattling down some code. I even used virtual for all navigation properties without the intent of lazy loading.. ;)
